src/DataStreams/RemoteQueryExecutor.cpp

```diff
@@ -176,7 +186,37 @@ Block RemoteQueryExecutor::read()
         if (context.getSettingsRef().skip_unavailable_shards && (0 == multiplexed_connections->size()))
             return {};
     }
+    try
+    {
+        return readPackets();
+    }
+    catch (const DB::Exception & ex)
+    {
+        if (ex.code() == ErrorCodes::DUPLICATED_UUIDS)
```

Comment:
I have an awful failure case for this where it will never make any progress:

* 3 shard cluster
* shard 1 has part 1, shard 2 has part 2 and shard 3 has both part 1 and part 2 (2 concurrent part moves).
* the packet with part uuids is received first from shard 1
  * then from shard 3
    * now this query is resent as it contains part duplicates (in flight)
* packet with part uuids is received from shard 2
* packet with part uuids is received from shard 3 which again contains duplicates

One solution is multiple retries with a very awful worst case.

I'm ok with ignoring this case for now but it is worth being mentioned.

Reply:
And `max_distributed_connections=1` is a workaround for such case, right?